Add alerts to the sample Memcached operator

assafad opened the pull request:
Signed-off-by: assaf-admi <aadmi@redhat.com>

<!--

Welcome to the Operator SDK! Before contributing, make sure to:

- Read the contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD
- Rebase your branch on the latest upstream master
- Link any relevant issues, PR's, or documentation
- Check that the commit message is concice and helpful:
    - When fixing an issue, add "Closes #<ISSUE_NUMBER>"
    - Sign your commit https://github.com/apps/dco
- Follow the below checklist if making a user-facing change 

-->

**Description of the change:**
This PR adds alerts and recording rule to the monitoring sample Memcached operator, including unit tests and runbooks for the new alerts.
Added alerts:
```
MemcachedDeploymentSizeUndesired
MemcachedOperatorDown
```
Added recording rule:
```
memcached_operator_up_total
```
In addition, this PR is addressing https://github.com/operator-framework/operator-sdk/issues/6140 as agreed.

**Motivation for the change:**
We would like to implement observability concepts for the Memcahced operator, in order to have the operator in [Deep Insights](https://sdk.operatorframework.io/docs/overview/operator-capabilities/#level-4---deep-insights) capability level. Then, we may use it as a reference in an observability best-practices tutorial in Operator-SDK. 

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

assafad commented:
Hi @machadovilaca, @sradco, this is an initial PR for the sample Memcached operator alerts, in order to hear your thoughts on the alerts, recording rule and unit tests that were added. It's probably missing a few steps before merging (e.g. not generated with `make generate` yet), and as @varshaprasad96 suggested, it may take place only in a separate monitoring sample operator.

machadovilaca commented:
I like it, I think it looks good so far

sradco commented on testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml:
```diff
@@ -0,0 +1,62 @@
+group_eval_order:
+- memcached.rules
+
+tests:
```
@assafad Hi, Please look at this example https://github.com/kubevirt/kubevirt/blob/11316e05586b95dd3b3b4898f19cb2a743031e2c/hack/prom-rule-ci/prom-rules-tests.yaml#L71.

You can add additional values and test the result in different eval_time and merge some of the tests together.

assafad replied:
Done

sradco commented on testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml:
```diff
@@ -0,0 +1,62 @@
+              description: "No running memcached-operator pods were detected in the last 5 min."
+            exp_labels:
+              severity: "critical"
+
```
@assafad Please remove empty lines

assafad replied:
Done

sradco commented on testdata/go/v4-alpha/memcached-operator/monitoring/alerts.go:
```diff
@@ -0,0 +1,81 @@
+		},
+		Labels: map[string]string{
+			"severity": "warning",
+		},
```
@assafad I think we should add example runbooks to the alerts.

assafad replied:
Done

sradco commented on testdata/go/v4-alpha/memcached-operator/go.sum:
```diff
@@ -198,8 +198,8 @@ github.com/google/go-cmp v0.5.6/go.mod h1:v8dTdLbMG2kIc/vJvl+f65V22dbkXbowE6jgT/
-github.com/google/gofuzz v1.1.0 h1:Hsa8mG0dQ46ij8Sl2AYJDUv1oA9/d6Vk+3LG99Oe02g=
-github.com/google/gofuzz v1.1.0/go.mod h1:dBl0BpW6vV/+mYPU4Po3pmUjxk6FQPldtuIdl/M65Eg=
+github.com/google/gofuzz v1.2.0 h1:xRy4A+RhZaiKjJ1bPfwQ8sedCA+YS2YcCHW6ec7JMi0=
+github.com/google/gofuzz v1.2.0/go.mod h1:dBl0BpW6vV/+mYPU4Po3pmUjxk6FQPldtuIdl/M65Eg=
```
@assafad I think we should remove these changes and only add the lines related to
```
github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.60.1
```

assafad replied:
I believe these are all dependencies. I got an error about them while `docker build & push`

sradco commented on testdata/go/v4-alpha/memcached-operator/go.sum:
```diff
@@ -436,8 +438,8 @@ golang.org/x/net v0.0.0-20210405180319-a5a99cb37ef4/go.mod h1:p54w0d4576C0XHj96b
-golang.org/x/net v0.0.0-20220722155237-a158d28d115b h1:PxfKdU9lEEDYjdIzOtC4qFWgkU2rGHdKlKowJSMN9h0=
-golang.org/x/net v0.0.0-20220722155237-a158d28d115b/go.mod h1:XRhObCWvk6IyKnWLug+ECip1KBveYUHfp+8e9klMJ9c=
+golang.org/x/net v0.0.0-20220826154423-83b083e8dc8b h1:ZmngSVLe/wycRns9MKikG9OWIEjGcGAkacif7oYQaUY=
+golang.org/x/net v0.0.0-20220826154423-83b083e8dc8b/go.mod h1:YDH+HFinaLZZlnHAfSS6ZXJJ9M9t4Dl22yv3iI2vPwk=
```
@assafad same comment as above

sradco commented on testdata/go/v4-alpha/memcached-operator/go.mod:
```diff
@@ -5,6 +5,7 @@ go 1.19
 require (
 	github.com/onsi/ginkgo/v2 v2.1.4
 	github.com/onsi/gomega v1.19.0
+	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.60.1
```
@assafad I think we should only add this change and remove changes that are not related to it, unless they are dependencies of 
```
github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.60.1
```

assafad commented:
/hold

assafad commented:
Hi @sradco, @machadovilaca, I updated the PR with a few changes:
- Added alerts runbooks
- Modified [prom-rules-tests.yaml](https://github.com/operator-framework/operator-sdk/pull/6135/files#diff-7f2eaf6ca58d6145dfbadcea55f3565d48538734770ca81ea2fce9c719d44cde)
- Since we now have a sample operator with default monitoring, I added to it some default RBAC in order to give Prometheus access to the operator's alerts. It assumes using kube-prometheus, as recommended in [exporting-metrics-for-prometheus](https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus). The alerts tutorial will mention that, and will explain how to change it.
- Modified [memcached_controller.go](https://github.com/operator-framework/operator-sdk/pull/6135/files#diff-ecc9df479abf7d65e51c0ffdb46ec14c35055978bd85cc52c8f4b6f473207b52) to create PrometheusRule only when it doesn't exist yet.

Please have another look. One you will approve the changes, I will work to have it all scaffold.

assafad commented:
Hi @varshaprasad96, @everettraven, could you please have a look?
This PR doesn't yet include changes in the samples generator, but only the suggested changes for the monitoring sample operator. Once the changes will be approved, I will work to have it scaffold.

everettraven commented on testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go:
```diff
@@ -237,7 +237,12 @@ var _ = Describe("memcached", Ordered, func() {
 			finalMetricValue := getMetricValue(memcachedDeploymentSizeUndesiredCountTotalName)
-			Expect(finalMetricValue).To(Equal(initialMetricValue + numberOfScales))
+
+			isFinalValueGreater := false
+			if finalMetricValue >= initialMetricValue+numberOfScales {
+				isFinalValueGreater = true
+			}
+			Expect(isFinalValueGreater).To(BeTrue())
```
Since the tests use `gomega`,  I *think* this could be simplified to one line with:
```go
Expect(1.0).Should(BeNumerically(">=", 1.0))
```
For more info on the `BeNumerically` function you can take a look at the [`gomega` go docs entry](https://pkg.go.dev/github.com/onsi/gomega#BeNumerically)

assafad replied:
Done

everettraven reviewed (approved):
@assafad Thanks for the contribution! Overall it looks good to me.

I have one tiny non-blocking nit:

machadovilaca commented on testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go:
```diff
@@ -83,11 +88,23 @@ type MemcachedReconciler struct {
 func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := log.FromContext(ctx)
 
+	// Check if prometheus rule already exists, if not create a new one
```
not only you should check if the rule exists, but also if it was externally changed, and if so revert it to the initial configuration

assafad replied:
I added it, please let me know the change makes sense.

machadovilaca replied:
looks good

machadovilaca commented on testdata/go/v4-alpha/monitoring/memcached-operator/docs/runbooks/MemcachedDeploymentSizeUndesired.md:
```diff
@@ -0,0 +1,33 @@
+# MemcachedDeploymentSizeUndesired
+
+## Meaning
+MemcachedDeploymentSizeUndesired is triggered when there is a steep increase in the value of `memcached_deployment_size_undesired_count_total` counter metric. 
```
I think here it makes more sense to explain why the alert was triggered. More like "MemcachedDeploymentSizeUndesired is triggered when the number of available replicas of the Memcached deployment doesn't match the requested configuration"

assafad replied:
Thanks, done

machadovilaca commented on testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:
```diff
@@ -0,0 +1,84 @@
+	deploymentSizeUndesiredAlert = "MemcachedDeploymentSizeUndesired"
+	operatorDownAlert            = "MemcachedOperatorDown"
+	operatorUpTotalRecordingRule = "memcached_operator_up_total"
+	runbookURLBasePath           = "https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v4-alpha/monitoring/memcached-operator/runbooks/"
```
Will this be hardcoded in operator-sdk generator or are you thinking about configuring it somehow?

assafad replied:
@machadovilaca not sure I understood exactly what you are referring. Please explain more if possible.

machadovilaca replied:
every new operator created will have a different runbookURLBasePath, right? My question was if you intended to leave this as `https://github.com/operator-framework/operator-sdk/tree...` by default and let the user change that later, or if you create it on operator code generation time

assafad replied:
Not sure what it has to change. The samples are stable on Github, no?
This code is only for the monitoring operator sample which will always have `memcached-operator` name, and its runbook base URL will always be `https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v4-alpha/monitoring/memcached-operator/runbooks/`. 
But please correct me if I'm wrong.

assafad replied:
> every new operator created will have a different runbookURLBasePath

Why is that?

machadovilaca replied:
in testdata yes, it won't change, my question is more for when you do the generator.

assafad replied:
the generator is also for the samples, so don't see why this URL should be generated differently than other parts.
Basically, I planned to generate it hardcoded

machadovilaca replied:
that generator is what operator-sdk uses to create new operator projects. What will those operators have in `runbookURLBasePath`?

assafad replied:
AFAIK, the generator we're working on is only for the samples on testdata, which are not scafold by the default  operator-sdk scaffolding, no? 
https://github.com/operator-framework/operator-sdk/blob/master/hack/generate/samples/generate_testdata.go#L64

machadovilaca replied:
mhh, maybe, but then users won't have the monitoring stuff if they generate a new project, right?

everettraven replied:
@assafad @machadovilaca So my understanding on this PR is specifically to add the alerts and corresponding code *only* on the testdata. We have a generator specifically for generating our testdata that @assafad is updating and not the actual generation logic used in the `operator-sdk` cli tool that users would use to scaffold their projects.

That being said, I think @machadovilaca points do raise the question - If the plan is to ultimately create a plugin that will do most of the heavy lifting by scaffolding most of this logic, should we really be updating the testdata prior to creating that plugin? It feels to me like it may be better to go ahead and start writing the plugin and tackle updating the testdata once the plugin is written so you aren't writing testdata generation logic just to delete it later when the plugin automatically scaffolds *most* of that logic for you.

machadovilaca replied:
we actually just met to discuss this, and we agree with you, it does not make sense to update the testdata generator, but to first create a plugin in kubebuilder instead

machadovilaca commented:
would also like to see a test to verify all alerts have a runbook url

sradco commented:
> > would also like to see a test to verify all alerts have a runbook url
> 
> @machadovilaca I plan to have another PR with e2e testing, which includes alerts test and rules tests (verifying each alert has `summary`, `severity`, accessible `runbook_url`).

runbook_url is optional , so we will not require it in the alert fields test.

varshaprasad96 commented:
@assafad The contents added to the sample project looks good. What I am concerned about is that these are in the testdata/ directly, instead they need to be in the generator or else re-generating the testdata will wipe these changes. Another addition which will be really useful is to add documentation in the SDK website (https://sdk.operatorframework.io), about this sample monitoring project, the additions made and benefits for anyone who is looking to add a prometheus monitoring stack.

assafad commented:
> @assafad The contents added to the sample project looks good. What I am concerned about is that these are in the testdata/ directly, instead they need to be in the generator or else re-generating the testdata will wipe these changes. Another addition which will be really useful is to add documentation in the SDK website (https://sdk.operatorframework.io), about this sample monitoring project, the additions made and benefits for anyone who is looking to add a prometheus monitoring stack.

@varshaprasad96 thanks. I'm now working on generating the changes which are proposed. Regarding your suggestion for documentation, we have an [issue](https://issues.redhat.com/browse/CNV-21038) for that, and will address it in a different PR.

varshaprasad96 commented on testdata/go/v4-alpha/monitoring/memcached-operator/go.mod:
```diff
@@ -3,13 +3,14 @@ module github.com/example/memcached-operator
-	k8s.io/apimachinery v0.25.0
-	k8s.io/client-go v0.25.0
-	sigs.k8s.io/controller-runtime v0.13.0
+	k8s.io/api v0.25.4
```
Can we avoid bumping k8s version in this PR? This is because the k8s bumps (minor or major) are usually tracked separately so as changes happening due to them.

assafad replied:
I'm not sure. Maybe it's running `go mod tidy` after generating? How can I avoid these bumps?

varshaprasad96 commented on testdata/go/v4-alpha/monitoring/memcached-operator/go.mod:
```diff
@@ -3,13 +3,14 @@ module github.com/example/memcached-operator
+	k8s.io/api v0.25.4
+	k8s.io/apimachinery v0.25.4
+	k8s.io/client-go v0.25.4
+	sigs.k8s.io/controller-runtime v0.13.1
```
Same with controller-runtime

assafad commented on testdata/go/v3/monitoring/memcached-operator/monitoring/prom-rule-ci/prom-rule-tests.yaml:
```diff
@@ -0,0 +1,66 @@
+---
```
Added more explanations and a link to Prometheus official documentation

avlitman commented on hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:
```diff
@@ -568,7 +763,26 @@ func (mh *Memcached) customizingDockerfile() {
 
+// customizingMakefile will customize the Makefile to include monitoring
+func (mh *Memcached) customizingMakefile() {
+	dockerfilePath := filepath.Join(mh.ctx.Dir, "Makefile")
```
```suggestion
	makefilePath := filepath.Join(mh.ctx.Dir, "Makefile")
```

assafad replied:
thanks, done

avlitman commented on hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:
```diff
@@ -568,7 +763,26 @@ func (mh *Memcached) customizingDockerfile() {
+	dockerfilePath := filepath.Join(mh.ctx.Dir, "Makefile")
+
+	// Add prom-rule-ci target
+	err := kbutil.InsertCode(dockerfilePath,
```
```suggestion
	err := kbutil.InsertCode(makefilePath,
```

assafad replied:
thanks, done

avlitman commented on hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:
```diff
@@ -513,46 +524,230 @@ func (mh *Memcached) implementingAPIMarkers() {
+func (mh *Memcached) implementingMetrics() {
 	// Create metrics file
-	_, err = os.Create(filepath.Join(mh.ctx.Dir, "monitoring/metrics.go"))
+	_, err := os.Create(filepath.Join(mh.ctx.Dir, "monitoring/metrics.go"))
 	pkg.CheckError("creating metrics file", err)
 
 	metricsPath := filepath.Join(mh.ctx.Dir, "monitoring/metrics.go")
```
```suggestion
	metricsPath := filepath.Join(mh.ctx.Dir, "monitoring/metrics.go")
	_, err := os.Create(metricsPath)
	pkg.CheckError("creating metrics file", err)

```

avlitman replied:
You use `filepath.Join(mh.ctx.Dir, "monitoring/metrics.go")` twice, no need (Joao comment this on my pr as well)

assafad replied:
Done for all created files, thanks!

avlitman commented on hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:
```diff
@@ -513,46 +524,230 @@ func (mh *Memcached) implementingAPIMarkers() {
 
+func (mh *Memcached) implementingAlerts() {
+	// Create alerts file
+	_, err := os.Create(filepath.Join(mh.ctx.Dir, "monitoring/alerts.go"))
+	pkg.CheckError("creating alerts file", err)
+
+	alertsPath := filepath.Join(mh.ctx.Dir, "monitoring/alerts.go")
```
```suggestion
	alertsPath := filepath.Join(mh.ctx.Dir, "monitoring/alerts.go")
	_, err := os.Create(alertsPath)
	pkg.CheckError("creating alerts file", err)

```

assafad replied:
Done for all created files, thanks!

varshaprasad96 commented on testdata/go/v3/monitoring/memcached-operator/controllers/memcached_controller.go:
```diff
@@ -36,11 +37,15 @@ import (
 
 const memcachedFinalizer = "cache.example.com/finalizer"
+const ruleName = "memcached-operator-rules"
+const namespace = "memcached-operator-system"
```
can we get this from `req.NamespacedName`? This contains the name and namespace of the object that would be reconciled. Should the `monitoringv1.PrometheusRule` be in the same location?

assafad replied:
I wasn't able to get this from `req.NamespacedName` which AFAIK refers to `memcached` here. 
Yes, we create the `monitoringv1.PrometheusRule` in the same namespace.

assafad commented:
/retest

varshaprasad96 commented:
The PR looks good to me! Thanks for the contribution @assafad!
Waiting for another set of eyes to review before merging. cc: @everettraven @camilamacedo86 @Kavinjsir @theishshah
